files/image_config/corefile_uploader/core_uploader.py

```diff
@@ -0,0 +1,276 @@
+
+class Handler(FileSystemEventHandler):
+
+    @staticmethod
+    def init():
+        global hostname, sonicversion, asicname, acctname, acctkey, sharename
+        global cwd, cfg
+
+        cfg = config()
+
+        set_env(cfg.get_dict()["env"])
+
+        hostname = socket.gethostname()
+        if not hostname:
+            raise Exception("Failed to read hostname")
+
+        acctname = cfg.get_data(("azure_sonic_core_storage", "account_name"))
+        acctkey = cfg.get_data(("azure_sonic_core_storage", "account_key"))
+        sharename = cfg.get_data(("azure_sonic_core_storage", "share_name"))
+
+        if not acctname or not acctkey or not sharename:
+            while True:
+                # Wait here until service restart
+                log_err("Unable to retrieve Azure storage credentials")
+                time.sleep (HOURS_4)
+
+        with open("/etc/sonic/sonic_version.yml", 'r') as stream:
+            l = yaml.safe_load(stream)
+            sonicversion = l['build_version']
+            asicname = l['asic_type']
+
+        if not sonicversion:
+            raise Exception("Failed to read build_version from /etc/sonic/sonic_version.yml")
+
+        if not asicname:
+            raise Exception("Failed to read asic_type from /etc/sonic/sonic_version.yml")
+
+        cwd = cfg.get_data(("local_work", "core_upload")).split("/")
+        if not len(cwd) > 2:
+            raise Exception("Invalid path for core_upload. Expect a min of two elements in path")
+
+        os.chdir(INIT_CWD)
+
+    @staticmethod
+    def on_any_event(event):
+        if event.is_directory:
+            return None
+
+        elif event.event_type == 'created':
+            # Take any action here when a file is first created.
+            log_debug("Received create event - " +  event.src_path)
+            Handler.handle_file(event.src_path)
+
+
+    @staticmethod
+    def wait_for_file_write_complete(path):
+        ct_size = -1
+
+        while ct_size != os.path.getsize(path):
+            ct_size = os.path.getsize(path)
+            time.sleep(2)
+
+        time.sleep(2)
+        if ct_size != os.path.getsize(path):
+            raise Exception("Dump file creation is too slow: " + path)
+
+        log_debug("File write complete - " +  path)
+
+
+    @staticmethod
+    def handle_file(path):
+
+        Handler.wait_for_file_write_complete(path)
+
+        lpath = "/".join(cwd)
+        make_new_dir(lpath)
+        os.chdir(lpath)
+
+        # Create a new archive with core & more.
+        metafiles = cfg.get_dict()["metadata_files_in_archive"]
+
+        fname = os.path.basename(path)
+        tarf_name = fname + ".tar.gz"
+
+        cfg.get_core_info(path, hostname)
+        
+        tar = tarfile.open(tarf_name, "w:gz")
+        for e in metafiles:
+            tar.add(metafiles[e])
+        tar.add(path)
+        tar.close()
+        log_debug("Tar file for upload created: " + tarf_name)
+
+        Handler.upload_file(tarf_name, tarf_name)
```

Comment:
is the core file removed after upload?

Reply:
No.
Reasons:

1) I am not attaching core file to ICM now, so as not to increase ICM DB size,. Hence we need it here.
2) With core files remaining, upon service restart, it will re upload, 
    a) Re-upload is not a problem, as the analyzer would understand that it is duplicate and will not file any ticket
    b) If the file is still in azure storage, it would only get re-updated and no duplicates.
In a normal system, the service restart should be pretty seldom (only upon reboot)

3) When s process dumps cores repeatedly, only one ticket is created. Even if we attach core file., we would attach only one. But often, we might like to see multiple dumps during core file analysis to help see a pattern.

On a side note: FYI:
5) We have an upcoming core-mgmt service from Broadcom that would limit the count of core files for a process